feat(eks): add deletionProtection property to Cluster construct

matoom-nomu opened the pull request:
### Issue # (if applicable)

Closes #36460

### Reason for this change

AWS EKS supports deletion protection to prevent accidental cluster deletion, as documented in the [CloudFormation reference](https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-resource-eks-cluster.html#cfn-eks-cluster-deletionprotection).

### Description of changes

- Add `deletionProtection` optional property to `ClusterProps` interface
- Update README with usage example

### Describe any new or updated permissions being added

None


### Description of how you validated changes

Added unit/integ tests.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

badmintoncryer commented on packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:
```diff
@@ -3962,3 +3962,45 @@ describe('cluster', () => {
+        deletionProtection: expected,
+      },
+    });
+  });
```
You can describe test conditions more simply.

```suggestion
  test.each([
    true, false
  ])('deletionProtection(%s) should work', (deletionProtection) => {
    // GIVEN
    const { stack } = testFixture();

    // WHEN
    new eks.Cluster(stack, 'Cluster', {
      version: CLUSTER_VERSION,
      deletionProtection,
      kubectlLayer: new KubectlV31Layer(stack, 'KubectlLayer'),
    });

    // THEN
    Template.fromStack(stack).hasResourceProperties('Custom::AWSCDK-EKS-Cluster', {
      Config: {
        deletionProtection,
      },
    });
  });
```

matoom-nomu replied:
I've simplified the test descriptions as you suggested.

badmintoncryer commented on packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:
```diff
@@ -3962,3 +3962,45 @@ describe('cluster', () => {
+    // THEN
+    Template.fromStack(stack).hasResourceProperties('Custom::AWSCDK-EKS-Cluster', {
+      Config: Match.not(Match.objectLike({
+        deletionProtection: Match.anyValue(),
```
Is it better to use `Match.absent()`?

badmintoncryer commented on packages/@aws-cdk-testing/framework-integ/test/aws-eks/test/integ.eks-cluster-deletion-protection.ts:
```diff
@@ -0,0 +1,52 @@
+  },
+});
+
+app.synth();
```
This line is not needed.

```suggestion
```

badmintoncryer commented on packages/aws-cdk-lib/aws-eks/lib/cluster.ts:
```diff
@@ -732,6 +732,15 @@ export interface ClusterOptions extends CommonClusterOptions {
+   * When true, deletion protection is enabled and the cluster cannot be deleted until protection is disabled.
+   * When false, the cluster can be deleted normally. This setting only applies to clusters in an active state.
+   *
+   * @default - none
```
Could you please describe the default value or  behavior that happens if this property is not supplied?
```suggestion
   // default value
   * @default false
   // default behaviour
   * @default - deletion protection is disabled
```

matoom-nomu commented on packages/@aws-cdk-testing/framework-integ/test/aws-eks/test/integ.eks-cluster-deletion-protection.ts:
```diff
@@ -0,0 +1,52 @@
+        rollback: true,
+      },
+    },
+  },
```
same as above

badmintoncryer commented:
Could you please resolve build error?

```console
aws-cdk-lib: /codebuild/output/src3139386751/src/actions-runner/_work/aws-cdk/aws-cdk/packages/aws-cdk-lib/aws-eks/lib/cluster.ts
aws-cdk-lib:   740:5  error  Trailing spaces not allowed  no-trailing-spaces
aws-cdk-lib:   742:5  error  Trailing spaces not allowed  no-trailing-spaces
aws-cdk-lib: /codebuild/output/src3139386751/src/actions-runner/_work/aws-cdk/aws-cdk/packages/aws-cdk-lib/aws-eks/test/cluster.test.ts
aws-cdk-lib:   3968:16  error  Missing trailing comma  @stylistic/comma-dangle
```

badmintoncryer commented:
@matoom-nomu Thanks! 

> Should i resolve security gurdian checks as well ?

It may be not essential. Security guardian CI is introduced experimentally and I think it has no effect to review process.

Could you please address this comment? I'll approve this PR after that.

https://github.com/aws/aws-cdk/pull/36474#discussion_r2653257896

Have a happy new year!!

badmintoncryer reviewed (approved):
Thanks!

matoom-nomu commented:
> @matoom-nomu Thanks!
> Could you please address this comment? I'll approve this PR after that.
> 
> [#36474 (comment)](https://github.com/aws/aws-cdk/pull/36474#discussion_r2653257896)
> 
> Have a happy new year!!

Thanks for the review, have a happy new year too!

And it seems there is a build error due to the year change....

>@aws-cdk/mixins-preview: - [license/notice-file] NOTICE does NOT begin with AWS Cloud Development Kit (AWS CDK),Copyright 2018-2026 Amazon.com, Inc. or its affiliates. All Rights Reserved.,'

https://github.com/aws/aws-cdk/blob/2845d47e074e3bf7e8d36bf655822c70ae53f094/packages/@aws-cdk/mixins-preview/NOTICE#L2

Shall i make the Issue ticket about this ?

badmintoncryer commented:
@matoom-nomu This is annual problem. How about submit PR like this?

https://github.com/aws/aws-cdk/pull/32705

matboros commented:
Hi,

thanks a lot for this PR. I see it has been sitting low for a while. I rebased it on the latest main, and the build is failing.

Apart from fixing that, my high level comments would be:

- my understanding is that the EKS Cluster is being managed by the custom resource in the `packages/@aws-cdk/custom-resource-handlers/lib/aws-eks/cluster-resource-handler/cluster.ts` file. So in order for the property to be properly supported, we would need to fix the following 2 things:
  - The `parseProps` function converts other booleans (like `endpointPrivateAccess`) from string to boolean, but doesn't handle `deletionProtection`. CloudFormation passes custom resource props as strings, so the EKS SDK will reject "true" instead of true.
  - The `analyzeUpdate` function and `UpdateMap` don't include `deletionProtection`. Changing this value on an existing cluster will silently do nothing. The EKS API supports updating via UpdateClusterConfig.
- In the integration test, creating a cluster with `deletionProtection: true` means the stack can't be destroyed. The `onDelete` handler doesn't disable protection before deleting. A nice way would be to disable the protection on the construct, then regenerate the CFN template, and deploy it on `onDelete`. This would then additionally test that the update path works as well.

What do you think?

matoom-nomu commented:
@matboros 
Thanks for the detailed review. I'll push the updates to this PR shortly.

matoom-nomu commented:
@matboros 
I've addressed all three points you mentioned.

1. **parseProps** 
Added string-to-boolean conversion for `deletionProtection`.

2. **analyzeUpdate / UpdateMap**
Added `updateDeletionProtection` to detect changes and call `updateClusterConfig` accordingly.

3. **integTest** 
I followed the pattern from #36924 (ASG deletion protection): the integ test verifies DELETE_FAILED, then updates the stack with protection disabled for cleanup. 

could you review this again?

matboros commented on packages/@aws-cdk/custom-resource-handlers/lib/aws-eks/cluster-resource-handler/cluster.ts:
I have a concern with this line. Here we actually say, that updateDeletionProtection can be done alongside any other change, but when there is a version change, the method returns early, and the deletion protection is not updated.

We could either move the deletion protection update to the beginning, like the tag update, but then you add the complexity to run it if the version is updated too, or just treat it like all other updates, so they cannot appear alongside the version update, or other updates. The second seems easier to me. I understand it is more restrictive, but aligns with the error, and since this is a new feature anyways, we can see if there is significant demand on supporting multiupdate (I doubt, since ppl should migrate to EKS v2 anyways if they can).

matoom-nomu replied:
Thanks for review, went with option 2. updateDeletionProtection is now treated like other updates and cannot appear alongside them.

matboros commented on packages/@aws-cdk-testing/framework-integ/test/aws-eks/test/integ.eks-cluster-deletion-protection.js.snapshot/aws-cdk-eks-cluster-deletion-protection.template.json:
```diff
@@ -788,7 +788,7 @@
      },
      "accessConfig": {},
-     "deletionProtection": true
+     "deletionProtection": false
```
shouldnt this be true?

matoom-nomu replied:
yeah this should be `true` , Modified integ-test,and confrimed it's now `true`.

matboros commented on packages/@aws-cdk-testing/framework-integ/test/aws-eks/test/integ.eks-cluster-deletion-protection.ts:
```diff
@@ -0,0 +1,81 @@
+    super(scope, id, props);
+
+    // allow all account users to assume this role in order to admin the cluster
+    const mastersRole = new iam.Role(this, 'AdminRole', {
```
This role is flagged  by the security guardian. since this is in all other tests too, lets just suppress that for now with:
```
  // Suppress Security Guardian: this role intentionally uses account root principal
  // so any account user can administer the cluster for testing purposes
  (mastersRole.node.defaultChild as CfnResource).addMetadata('guard', {
    SuppressedRules: ['IAM_ROLE_ROOT_PRINCIPAL_NEEDS_CONDITIONS'],
  });
```

matoom-nomu replied:
Done, modified as you mentioned.

matboros commented on packages/@aws-cdk/custom-resource-handlers/lib/aws-eks/cluster-resource-handler/cluster.ts:
```suggestion
      name: clusterName,
      // deletionProtection is not yet in the SDK types, so we need to explicitly
      // pass it to prevent SDK v3 from stripping the unknown property during serialization.
      deletionProtection: (this.newProps as any).deletionProtection,
```

matboros replied:
I ran the integ tests with deployment, and the cluster was deletable because the flag was not applied apparently because of sdk stripping unknown fields. With this change the test are passing

matoom-nomu replied:
@matboros 
Thanks, fixed. 
I confirmed integ test worked with the code you suggested.
Sorry for the back and forth on this one!
